keps/sig-api-machinery/5000-api-linting-crd-schema-tooling/README.md

```diff
@@ -0,0 +1,522 @@
+- Validating changes to CustomResourceDefinition schemas in YAML
+
+[kube-api-conventions]:(https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+[making-changes-to-apis]:(https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md)
+
+### Goals
+
+- Create tooling to help developers writing Kubernetes APIs in Go to follow best practices and Kubernetes API conventions
+- Create tooling to help developers writing CustomResourceDefinitions avoid making breaking changes
+
+### Non-Goals
+
+- Create tooling for validating CustomResourceDefinition schema changes at admission time
+    - This is a non-goal of this KEP in particular to keep the scope of this KEP to only include the individual tooling. A separate KEP may be written, focusing on addressing this goal.
+
+## Proposal
+
+This KEP introduces two new sig-api-machinery subprojects.
+The projects are intended to complement each other, and take lessons learned from [Kubernetes API conventions][kube-api-conventions] and our experience
+with [making changes to Kubernetes APIs][making-changes-to-apis], to enable the wider Kubernetes ecosystem to follow best practices and avoid breaking changes.
+
+### Go-based Kubernetes API Linting
+
+A linter, built as a `golangci-lint` plugin, that evaluates Go type defintions (typically `_types.go`), and flags deviations from Kubernetes API conventions and best practices.
+
+The linter will be configurable, allowing users to choose which of the rules they wish to adhere to, and which they wish to ignore.
+
+The linter will focus, as much as possible on providing a "built-ins" experience by default, targetting rules for types in Kubernetes/Kubernetes.
+It will also provide configuration to allow for a more "CRD" experience, adjusting the configuration and rules to cater to the needs of CRD authors.
+
+The linter will focus on static evaluation of CRD types.
+It cannot, and will not detect changes to the code, and therefore cannot assert any rule that requires information about a previous version of the types.
+
+### CRD Schema Change Validation
+
+Create a CLI tool that evaluates the differences between an old and new CRD in YAML to identify changes that may break users.
+
+All validations are exported in Go packages that can be imported and consumed by other projects.
+
+Some examples of how the CLI may be used:
+
+- Comparing CRD that exists on cluster to a local YAML file of the same CRD - `crdify kube://mycrd.example.org file://mycrd.yaml`
+- Comparing CRD across git revisions (could be used in CI systems) - `crdify git://main?path=mycrd.yaml git://HEAD?path=mycrd.yaml`
+- Comparing CRD from git revision to a local file - `crdify git://main?path=mycrd.yaml file://mycrd.yaml`
+
+The schema change validation will focus on changes that may break users, and changes that can be detected by comparing the CRD schemas in YAML.
```

Comment:
So this tool focused on backward compatibility and the linter focuses on conventions/best practices?

Reply:
Yeah that's correct. crdify will have a view of what's changed over time so is better suited to picking up things like "you shouldn't be increasing the minimum value".

We did actually start with our version of this tool doing both, but the feedback loop for developers was long. They would typically not run the schema diff checked until they created a PR, and of course, have to regenerate the CRDs to be able to even run this.

The linter however is much more lightweight, they don't have to generate the CRDs and it can even be integrated into IDEs to give users very quick feedback on the changes they've made.

I had suggested in the initial pass, we try and make them focused on the two separate "phases" of development, where each is best suited to check specific things